### DISCUSSION TITLE
DPGEN with constrained structures

### DISCUSSION THREAD

**Discussion by longkunxuluke** [deepmodeling/dpgen#877] [General]
Hi, thanks for developing this great tool. I am trying to use DPGEN to model battery interface where I am particularly interested in fixed a few layers of the electrode slab in the whole process (AIMD, DPGEN, and DPMD). I did not found relevant keywords in param.json. Is there any way to do this? Thanks

**Comment by wanghan-iapcm** [comment]
You may provide a template of lammps input to run the MD simulation as what you want. 
https://github.com/deepmodeling/dpgen/tree/master/examples/run/dp-lammps-enhance_sampling

**Comment by longkunxuluke** [comment]
Thanks for the reply. I am in the very first step, i.e., dpgen init_bulk PARAM [MACHINE], following is the structure of my Li slab where I intend to fix the bottom two layers (i.e., atom 33, 36, 39, 1, 17, 25, 5, 2, 18, 26, 6) in preparing the initial data. I am wondering if I can do this?

41
Li2
Li    0.000000    0.000000    0.000000
Li    0.000000    0.000000    3.426818
Li    0.000000   10.280454    0.000000
Li    0.000000   10.280454    3.426818
Li   10.280454    0.000000    0.000000
Li   10.280454    0.000000    3.426818
Li   10.280454   10.280454    0.000000
Li   10.280454   10.280454    3.426818
Li    0.000000    3.426818    0.000000
Li    0.000000    3.426818    3.426818
Li   10.280454    3.426818    0.000000
Li   10.280454    3.426818    3.426818
Li    0.000000    6.853636    0.000000
Li    0.000000    6.853636    3.426818
Li   10.280454    6.853636    0.000000
Li   10.280454    6.853636    3.426818
Li    3.426818    0.000000    0.000000
Li    3.426818    0.000000    3.426818
Li    3.426818   10.280454    0.000000
Li    3.426818   10.280454    3.426818
Li    3.426818    3.426818    0.000000
Li    3.426818    3.426818    3.426818
Li    3.426818    6.853636    0.000000
Li    3.426818    6.853636    3.426818
Li    6.853636    0.000000    0.000000
Li    6.853636    0.000000    3.426818
Li    6.853636   10.280454    0.000000
Li    6.853636   10.280454    3.426818
Li    6.853636    3.426818    0.000000
Li    6.853636    3.426818    3.426818
Li    6.853636    6.853636    0.000000
Li    6.853636    6.853636    3.426818
Li    1.713409    1.713409    1.713409
Li    1.713409    5.140227    1.713409
Li    1.713409    8.567045    1.713409
Li    5.140227    1.713409    1.713409
Li    5.140227    5.140227    1.713409
Li    5.140227    8.567045    1.713409
Li    8.567045    1.713409    1.713409
Li    8.567045    5.140227    1.713409
Li    8.567045    8.567045    1.713409

**Comment by HuangJiameng** [comment]
You can set `Selective dynamics` in POSCAR to fix the layers. https://wap.sciencenet.cn/blog-478347-1259200.html?mobile=1 I find a tutorial in Chinese. Maybe it can help you.

**Comment by AnguseZhang** [comment]
Previous solutions should work. It seems that this is more suitable as the discussion.

**Comment by taipinghu** [comment]
According to our experience, fix atoms in exploring stage is not a good solution because it can generate a large force and is bad for dpgen convergence.  
There is no need to fix atoms in the exploration step. You can add the bulk structures into the dataset. After collecting all data, There is also no need to fix atoms in the final DeePMD simulation because you can use the large cell.
